src/main/scala/metadata.scala

```diff
@@ -0,0 +1,83 @@
+package suffuse
+
+/** AKey is an attribute key: a Metadata map key which carries
+ *  both the type of the corresponding value, and a value to
+ *  provide when there is no value in the map. This as opposed
+ *  to wrapping everything in sight in Option.
+ */
+final class AKey[A](description: String, empty: => A) {
+  def emptyValue: A = empty
+  override def toString = description
+}
+
+/** Attr is a dependently typed key/value pair.
+ */
+trait Attr extends Any {
+  type Type
+  def key: AKey[Type]
+  def value: Type
+
+  def pair                               = key -> value
+  def matches[A](implicit that: AKey[A]) = key == that
+  def to_s                               = s"$key: $value"
+}
+
+/** The simple implementation of Attr, which ties the knot between the
+ *  user-facing type parameter and the type member which couples key and value.
+ */
+final case class AttrOf[A](key: AKey[A], value: A) extends Attr {
+  type Type = A
+  override def toString = s"$key: $value"
+}
+
+object Attr {
+  def apply[A](value: A)(implicit key: AKey[A]): AttrOf[A] = new AttrOf[A](key, value)
+  def unapply(x: Attr): Some[(AKey[x.Type], x.Type)]       = Some(x.pair)
+}
+
+/** A Metadata map, holding any number of well typed Attrs.
+ *  A typed value can be obtained for any key.
+ */
+final class Metadata(attrs: Vector[Attr]) {
+  private[this] val untypedMap                   = attrs.foldLeft(Map[AKey[_], Any]())(_ + _.pair)
+  private def doApply[A]()(implicit z: AKey[A]): A = untypedMap(z).asInstanceOf[A]
+
+  def isEmpty                                = untypedMap.isEmpty
+  def apply[A]()(implicit z: AKey[A]): A     = if (has[A]) doApply[A]() else z.emptyValue
```

Comment:
I didn't want to reproduce the entirety of psp-std here, but in case you haven't looked, maybe you'll like the way it's done there better.
```scala
/** The original type class for providing the "empty" value of a particular type.
 *  Suitable only for types with a unique (useful) definition of empty - but that's
 *  a lot of types. You could easily recover the behavior of methods like Option.get
 *  or Seq.head by creating a default instance of Empty[A] which throws an exception.
 */
trait Empty[@fspec +A] extends Any {
  def empty: A
}
```
Just look around psp-std for the ways `Empty[A]` is put to use. It's literally the best thing I've come up with relative to the status quo.